apollo-runtime-kotlin/src/iosMain/kotlin/com/apollographql/apollo/network/ApolloHttpNetworkTransport.kt

```diff
@@ -0,0 +1,208 @@
+      setHTTPMethod("GET")
+      httpHeaders.forEach { (key, value) -> setValue(value, forHTTPHeaderField = key) }
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+    }
+  }
+
+  private fun GraphQLRequest.toHttpPostRequest(): NSMutableURLRequest {
+    return NSMutableURLRequest.requestWithURL(serverUrl).apply {
+      val buffer = Buffer()
+      JsonWriter.of(buffer)
+          .beginObject()
+          .name("operationName").value(operationName)
+          .name("query").value(document)
+          .name("variables").value(variables)
+          .endObject()
+          .close()
+      val postBody = buffer.readByteArray().toNSData()
+
+      setHTTPMethod("POST")
+      httpHeaders.forEach { (key, value) -> setValue(value, forHTTPHeaderField = key) }
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+      setHTTPBody(postBody)
+    }
+  }
+
+  private fun parse(
+      data: NSData?,
+      httpResponse: NSHTTPURLResponse,
+      error: NSError?
+  ): Result {
+    if (error != null) return Result.Failure(
+        ApolloException(
+            message = "Failed to execute GraphQL http network request",
+            error = ApolloError.Network,
+            cause = IOException(error.localizedDescription)
+        )
+    )
+
+    val statusCode = httpResponse.statusCode.toInt()
+    if (statusCode !in 200..299) return Result.Failure(
+        ApolloException(
+            message = "Http request failed with status code `$statusCode`",
+            error = ApolloError.Network
+        )
+    )
+
+    if (data == null) return Result.Failure(
+        ApolloException(
+            message = "Failed to parse GraphQL http network response: EOF",
+            error = ApolloError.Network
+        )
+    )
+
+    return Result.Success(data)
+  }
+
+  sealed class Result {
+    class Success(val data: NSData) : Result()
+    class Failure(val cause: ApolloException) : Result()
+  }
+
+  private fun ByteArray.toNSData(): NSData = NSMutableData().apply {
+    if (isEmpty()) return@apply
+    this@toNSData.usePinned {
+      appendBytes(it.addressOf(0), size.convert())
+    }
+  }
+
+  @Suppress("NAME_SHADOWING")
+  private fun Result.dispatchOnMain(producerRef: COpaquePointer) {
+    val producerWithResultRef = StableRef.create((producerRef to this).freeze())
+    dispatch_async_f(
+        queue = dispatch_get_main_queue(),
+        context = producerWithResultRef.asCPointer(),
+        work = staticCFunction { ref ->
+          val producerWithResultRef = ref!!.asStableRef<Pair<COpaquePointer, Result>>()
+          val (producerPointer, result) = producerWithResultRef.get()
+          producerWithResultRef.dispose()
+
+          val producerRef = producerPointer.asStableRef<ProducerScope<GraphQLResponse>>()
+          val producer = producerRef.get()
+          producerRef.dispose()
+
+          when (result) {
+            is Result.Success -> {
+              producer.offer(
```

Comment:
Again if `producer` is a property of `NetworkTransport`, it's worth looking into what happens if the instance gets deallocated before this gets called. This is where you see `[weak self]` sprinkled liberally around Swift code, I'm not entirely sure what the equivalent is with coroutines and flows.

Reply:
We are safe here as `producer ` is not part of `NetworkTransport`, I mean `NetworkTransport ` doesn't hold reference.

It's a flow that holds reference, that returned to the user when he calls `ApolloClient.query`. KN doesn't have such concept as weak ref, but even if we keep reference for the producer, it's ok flow cancelation should take care of cleaning up.

And any `collecting` is tied to the coroutine scope, so if user doesn't leak scope and proper cancel flow we are safe here.

The only I guess improvements we missing here is to check if producer still alive before offer:
```
producer.isActive
producer.offer(...
```

Reply:
> The only I guess improvements we missing here is to check if producer still alive before offer

Isn't this the case already a few lines above ? (https://github.com/apollographql/apollo-android/blob/d464cf4fb0450f1f39716e2c4a41730dcf26f49b/apollo-runtime-kotlin/src/iosMain/kotlin/com/apollographql/apollo/network/ApolloHttpNetworkTransport.kt#L204)

All in all, I think we should wrap all `.offer()` with `runCaching {}`. We've been hit by  https://github.com/Kotlin/kotlinx.coroutines/issues/974 before.

Reply:
>Isn't this the case already a few lines above 

I've just added this right away.

>All in all, I think we should wrap all .offer() with runCaching {}

Right, I was thinking that even checking for isActive won't enough as it could be RC. Will wrap into runCatching